database/bfgd/database_ext_test.go

```diff
@@ -1653,6 +1653,168 @@ func TestL2BtcFinalitiesByL2KeystoneNotPublishedHeight(t *testing.T) {
+		},
+	}
+
+	createBlocksWithNoChildren := func(ctx context.Context, count int, db bfgd.Database) []int64 {
+		heights := make([]int64, count)
+		for i := range count {
+			height := mathrand.Int64()
+			hash := make([]byte, 32)
+			if _, err := rand.Read(hash); err != nil {
+				t.Fatal(err)
+			}
+			header := make([]byte, 80)
+			if _, err := rand.Read(header); err != nil {
+				t.Fatal(err)
+			}
+
+			btcBlock := bfgd.BtcBlock{
+				Height: uint64(height),
+				Hash:   hash,
+				Header: header,
+			}
+
+			if err := db.BtcBlockInsert(ctx, &btcBlock); err != nil {
+				t.Fatal(err)
+			}
+
+			heights[i] = height
+		}
+
+		return heights
+	}
+
+	createBlocksWithChildren := func(ctx context.Context, count int, db bfgd.Database, avoidHeights []int64, overlapHeights []int64) []int64 {
+		var prevHash []byte
+		overlapHeightI := 0
+		heights := make([]int64, count)
+		for i := range count {
+			var height int64
+			for {
+				if overlapHeightI < len(overlapHeights) {
+					height = overlapHeights[overlapHeightI]
+					overlapHeightI++
+					break
+				}
+
+				height = mathrand.Int64()
+				if !slices.Contains(avoidHeights, height) {
+					break
+				}
+			}
+			hash := make([]byte, 32)
+			if _, err := rand.Read(hash); err != nil {
+				t.Fatal(err)
+			}
+			header := make([]byte, 80)
+			if _, err := rand.Read(header); err != nil {
+				t.Fatal(err)
+			}
+
+			if len(prevHash) > 0 {
+				for k := range 32 {
+					header[k+4] = prevHash[k]
+				}
+			}
+
+			btcBlock := bfgd.BtcBlock{
+				Height: uint64(height),
+				Hash:   hash,
+				Header: header,
+			}
+
+			if err := db.BtcBlockInsert(ctx, &btcBlock); err != nil {
+				t.Fatal(err)
+			}
+			prevHash = hash
+			heights[i] = height
+		}
+		return heights
+	}
+
+	for _, tti := range testTable {
+		t.Run(tti.name, func(t *testing.T) {
+			ctx, cancel := defaultTestContext()
+			defer cancel()
+
+			db, sdb, cleanup := createTestDB(ctx, t)
+			defer func() {
+				db.Close()
+				sdb.Close()
+				cleanup()
+			}()
+
+			var overlapHeights []int64
+			noChildrenHeights := createBlocksWithNoChildren(ctx, tti.numberToCreateWithNoChildren, db)
+
+			childrenHeights := createBlocksWithChildren(ctx, tti.numberToCreateWithChildren, db, nil, overlapHeights)
+
+			if tti.overlapCount > 0 {
+				overlapHeights = noChildrenHeights[:tti.overlapCount]
+				oldChildrenHeights := childrenHeights
+				for _, o := range oldChildrenHeights {
+					if !slices.Contains(overlapHeights, o) {
+						childrenHeights = append(childrenHeights, o)
+					}
+				}
+			}
+
+			heights, err := db.BtcBlocksHeightsWithNoChildren(ctx)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			toCmp := make([]uint64, len(noChildrenHeights)+1)
+			for i, c := range noChildrenHeights {
+				toCmp[i] = uint64(c)
+			}
+			toCmp[len(toCmp)-1] = uint64(childrenHeights[len(childrenHeights)-1])
+
+			slices.Sort(heights)
+			slices.Sort(toCmp)
+
+			// we return a nil slice if emtpy, change that here for deep.Equal
```

Comment:
There's a typo here, oops:
```suggestion
			// we return a nil slice if empty, change that here for deep.Equal
```